incubator/hnc/internal/pkg/selectors/selectors.go

```diff
@@ -18,29 +20,47 @@ func GetSelectorAnnotation(inst *unstructured.Unstructured) string {
 	return annot[api.AnnotationSelector]
 }
 
+func GetTreeSelectorAnnotation(inst *unstructured.Unstructured) string {
+	annot := inst.GetAnnotations()
+	return annot[api.AnnotationTreeSelector]
+}
+
 // GetTreeSelector is similar to a regular selector, except that it adds the LabelTreeDepthSuffix to every string
 // To transform a tree selector into a regular label selector, we follow these steps:
 // 1. get the treeSelector annotation if it exists
 // 2. convert the annotation string to a slice of strings seperated by comma, because user is allowed to put multiple selectors
 // 3. append the LabelTreeDepthSuffix to each of the treeSelector string
 // 4. combine them into a single string connected by comma
 func GetTreeSelector(inst *unstructured.Unstructured) (labels.Selector, error) {
-	annot := inst.GetAnnotations()
-	treeSelectorStr, ok := annot[api.AnnotationTreeSelector]
-	if !ok {
+	treeSelectorStr := GetTreeSelectorAnnotation(inst)
+	if treeSelectorStr == "" {
 		return nil, nil
 	}
+
+	errStrs := []string{}
 	strs := strings.Split(treeSelectorStr, ",")
 	selectorStr := ""
 	for i, str := range strs {
+		// check if it's a valid namespace name
+		if str[0] == '!' {
+			errStrs = validation.IsDNS1123Label(str[1:])
+		} else {
+			errStrs = validation.IsDNS1123Label(str)
+		}
+		if len(errStrs) != 0 {
+			return nil, errors.New(str + " is not a valid namespace name, " + strings.Join(errStrs, ". "))
```

Comment:
`str` here could include the exclamation mark, because you didn't actually strip it out on line 46. Did you want to include the exclamation mark here?

Reply:
done